### PR TITLE
 Feat/remove mistaken registration

### PR DESCRIPTION
**Pull request by yumoL**
### Short description

implement #875 
### DoD

I have:
- [x] added actual code.
- [x] produced clean code that passes ESLint.
- [x] code that actually does what it should.
- [ ] documented the code or added documentation to the wiki.
- [x] added or modified test(s).
- [x] test(s) that actually pass.

### CODE REVIEW

File: backend/server/controllers/users.js [hunk 1]
```diff
@@ -245,5 +245,44 @@ module.exports = {
...
+          }
+        })
+        if (!siToRemove) {
+          // the student instance to be removed doesn't belong to the logged in user
+          return res.status(400).send('The student instance can be removed only by the student himself')
+        }
+        if (siToRemove.validRegistration) {
+          return res.status(400).send('The student instance can be removed only when his validRegistration has been marked as false')
+        }
+        // destroy the code review where toReview is req.body.id separately because the onDelete="cascade" has not been set to toReview
```
Comment (mjaakko) [inline backend/server/controllers/users.js]: Maybe deletions should be cascading or is there a problem with that?

File: labtool2.0/src/components/pages/CoursePage/StudentInfo.js [hunk 2]
```diff
@@ -44,9 +44,17 @@ export const CoursePageStudentInfo = props => {
           </Card>
           {courseData.role === 'student' && !courseData.data.validRegistration && (
             <>
-              <Message negative>
+              <Message className="mistakenRegistration" negative>
                 <Message.Header>Your registration has been marked as mistaken</Message.Header>
                 <p>This means you likely registered on this course by mistake. If this is not the case, please contact the teacher of the course.</p>
+                <Popup
+                  content="You can remove yourself from the course if you truely registered on it by mistake."
```
Comment (mjaakko) [inline labtool2.0/src/components/pages/CoursePage/StudentInfo.js]:
```suggestion
                  content="You can remove yourself from the course if you registered by mistake."
```

File: backend/server/controllers/users.js [hunk 3]
```diff
@@ -245,5 +245,44 @@ module.exports = {
...
+    }
+    if (req.authenticated.success) {
+      try {
+        const userId = req.decoded.id
+        const siToRemove = await StudentInstance.findOne({
+          where: {
+            id: req.body.id,
+            userId
+          }
+        })
```
Comment (mjaakko) [inline backend/server/controllers/users.js]: Maybe teachers should also be able to remove mistaken registrations?

Comment (yumoL) [reply backend/server/controllers/users.js]: Sounds reasonable, I'll change it later

File: labtool2.0/src/components/pages/CoursePage/StudentInfo.js (hunk 2 above)
Comment (hisahi) [reply labtool2.0/src/components/pages/CoursePage/StudentInfo.js]: Seconding this

File: backend/server/controllers/users.js (hunk 1 above)
Comment (hisahi) [reply backend/server/controllers/users.js]: It's probably been left out by someone else, maybe that is fixable

File: backend/server/controllers/users.js [hunk 4]
```diff
@@ -245,5 +245,49 @@ module.exports = {
...
+        }
+
+        const teacher = await TeacherInstance.findOne({
+          where: {
+            userId: req.decoded.id,
+            courseInstanceId: siToRemove.courseInstanceId
+          }
+        })
+        if (siToRemove.userId !== userId && !teacher) {
+          return res.status(400).send('You must be the student himself or the teacher of the courrse to remove this student instance')
```
Comment (hisahi) [inline backend/server/controllers/users.js]: `courrse` => `course`, and full stop missing. This could also be a 403 instead of 400.

**Pull request merged**